arangod/Cluster/ClusterMethods.cpp

```diff
@@ -139,18 +139,30 @@ Future<Result> beginTransactionOnSomeLeaders(TransactionState& state,
   TRI_ASSERT(state.isCoordinator());
   TRI_ASSERT(!state.hasHint(transaction::Hints::Hint::SINGLE_OPERATION));
 
-  std::shared_ptr<ShardMap> shardMap = coll.shardIds();
   ClusterTrxMethods::SortedServersSet leaders{};
 
-  for (auto const& pair : shards) {
-    auto const& it = shardMap->find(pair.first);
-    if (it->second.empty()) {
-      return TRI_ERROR_CLUSTER_BACKEND_UNAVAILABLE;  // something is broken
+  if (state.options().allowDirtyReads) {
+    // In this case we do not always choose the leader, but take the
+    // choice stored in the TransactionState. We might hit some followers
+    // in this case, but this is the purpose of `allowDirtyReads`.
+    for (auto const& pair : shards) {
+      ServerID const& replica = state.whichReplica(pair.first);
+      if (!state.knowsServer(replica)) {
+        leaders.emplace(replica);
```

Comment:
`leaders.emplace(replica)` took a while for me to digest.
Can we rename the variable `leaders` here to `servers`, because it now my contain leaders as well as followers?

Reply:
OK.